hieroplatform/src/main/java/org/hiero/remoting/HieroServer.java

```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * Implementation of sketch() service in hiero.proto.
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public void sketch(final Command command,
+                       final StreamObserver<PartialResponse> responseObserver) {
+        if (!checkValidIdsIndex(command.getIdsIndex(), responseObserver)) {
+            return;
+        }
+        final byte[] bytes = command.getSerializedOp().toByteArray();
+        final SketchOperation sketchOp = SerializationUtils.deserialize(bytes);
+        Observable<PartialResult> observable = this.dataSets.get(command.getIdsIndex())
+                                                                      .sketch(sketchOp.sketch);
+        Subscription sub = observable.subscribe(new Subscriber<PartialResult>() {
+            @Override
+            public void onCompleted() {
+                responseObserver.onCompleted();
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                responseObserver.onError(throwable);
+            }
+
+            @Override
+            public void onNext(PartialResult pr) {
+                final OperationResponse<PartialResult> res =
+                        new OperationResponse<PartialResult>(pr);
+                final byte[] bytes = SerializationUtils.serialize(res);
+                responseObserver.onNext(PartialResponse.newBuilder()
+                                                .setSerializedOp(ByteString.copyFrom(bytes))
+                                                .build());
+            }
+        });
+        this.operationToObservable.put(sketchOp.id, sub);
+    }
+
+    /**
+     * Implementation of zip() service in hiero.proto.
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public void zip(final Command command, final StreamObserver<PartialResponse> responseObserver) {
+        final byte[] bytes = command.getSerializedOp().toByteArray();
+        final ZipOperation zipOp = SerializationUtils.deserialize(bytes);
+        if (!checkValidIdsIndex(command.getIdsIndex(), responseObserver)
+            || !checkValidIdsIndex(zipOp.datasetIndex, responseObserver)) {
+            return;
+        }
+        final IDataSet other = this.dataSets.get(zipOp.datasetIndex);
+        Observable<PartialResult<IDataSet>> observable = this.dataSets.get(command.getIdsIndex())
+                                                            .zip(other);
+        Subscription sub = observable.subscribe(new Subscriber<PartialResult<IDataSet>>() {
+            @Override
+            public void onCompleted() {
+                responseObserver.onCompleted();
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                responseObserver.onError(throwable);
+            }
+
+            @Override
+            public void onNext(PartialResult<IDataSet> pr) {
+                if (isUnsubscribed()) {
```

Comment:
Even here you should assume that you can get multiple responses, only one of which can have a non-null response.